### DISCUSSION TITLE
The best way to install an extra Python package

### DISCUSSION THREAD

**Discussion by alihossein** [apache/airflow#15517] [General] [answered]
Hi, I wanted to install Python packages inside airflow containers. I don't know which method is better, please help.

**Comment by potiuk** [comment, marked as answer]
You need to build your own image see: https://airflow.apache.org/docs/docker-stack/build.html#a-pypi-package-example

**Reply by makrushin-evgenii** [reply]
Is there a way to provide packages via docker-compose, without building new image?

**Reply by potiuk** [reply]
This is certainly not recommended for production, but in the next version of airflow there will be a method to do it (via _PIP_ADDITIONAL_REQUIREMENTS variable):  https://github.com/apache/airflow/blob/main/docs/docker-stack/entrypoint.rst#installing-additional-requirements.

note this:
> The options below are mostly used for quick testing the image - for example with quick-start docker-compose or when you want to perform a local test with new packages added. They are not supposed to be run in the production environment as they add additional overhead for execution of additional commands. Those options in production should be realized either as maintenance operations on the database or should be embedded in the custom image used (when you want to add new packages).

**Reply by makrushin-evgenii** [reply]
Thanks a lot for your answer. To summarize, now to install extra dependencies I need:
1. Extend the image from apache/airflow:2.0.2 [according to docs](https://airflow.apache.org/docs/docker-stack/build.html#a-pypi-package-example)
2. Provide it in [docker-compose](https://airflow.apache.org/docs/apache-airflow/stable/docker-compose.yaml) as `AIRFLOW_IMAGE_NAME` param
3. To redefine dependencies I also should rebuild the image

Did I get it right?

**Reply by potiuk** [reply]
Correct.  You can also add "build ," to your docker-compose file and place your Dockerfile in the current dir similarly to https://docs.docker.com/compose/gettingstarted/#step-3-define-services-in-a-compose-file and `docker-compose up` should build your image automatically (`docker compose build` will rebuild it). This is also not recommended for production, however it should be one step less.

**Comment by john-tanium** [comment]


**Reply by Taragolis** [reply]
Just rebuild (preferable way) or extend existed image.

**Reply by potiuk** [reply]
You need to build your image and run containers based on it.

I think - if you want to continue using containers and images - you should invest your time in learning how it works. There is nothing "Airflow" special in using containers, those are really basic thing about containers, images, docker that you need to understand if you want to be efficient in it and not take time of those who can help with airflow issues on solving basic docker questions. This time is better spent elsewhere. There are plenty of tutorials available - guick googling revealed this https://docker-curriculum.com/ (but there are video, hands on and other tutorials) and I suggest you spend some time on going through those materials and even trying things yourself. A bit of exercise and trying out things how they work go a long way when it comes to understanding what you are doing.
